File: top-100-liked/34-find-first-and-last-position-of-element-in-sorted-array.py
# 在排序数组中查找元素的第一个和最后一个位置
# https://leetcode.cn/problems/find-first-and-last-position-of-element-in-sorted-array/
from typing import List


def searchLeft(nums: List[int], target: int) -> int:
    left, right = 0, len(nums) - 1
    while left <= right:
        mid = (right + left) // 2
        if nums[mid] == target:
            if mid == 0 or nums[mid - 1] < target:
                return mid
            else:
                right = mid - 1
        elif nums[mid] > target:
            right = mid - 1
        else:
            left = mid + 1
    return -1


def searchRight(nums: List[int], target: int) -> int:
    left, right = 0, len(nums) - 1
    while left <= right:
        mid = (right + left) // 2
        if nums[mid] == target:
            if mid == len(nums) - 1 or nums[mid + 1] > target:
                return mid
            else:
                left = mid + 1
        elif nums[mid] > target:
            right = mid - 1
        else:
            left = mid + 1
    return -1


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        left, right = searchLeft(nums, target), searchRight(nums, target)
        return [left, right]


nums = [5, 7, 7, 8, 8, 10]
target = 8
print(Solution().searchRange(nums, target))

File: top-100-liked/34-find-first-and-last-position-of-element-in-sorted-array.ts
/**
 * 在排序数组中查找元素的第一个和最后一个位置
 * https://leetcode.cn/problems/find-first-and-last-position-of-element-in-sorted-array/
 */

function binarySearch(nums: number[], target: number, lower: boolean) {
  let l = 0,
    r = nums.length,
    ans = nums.length
  while (l <= r) {
    const mid = Math.floor((l + r) / 2)
    if (nums[mid] > target || (lower && nums[mid] >= target)) {
      r = mid - 1
      ans = mid
    } else {
      l = mid + 1
    }
  }
  return ans
}

export function searchRange(nums: number[], target: number): number[] {
  let ans = [-1, -1]
  const l = binarySearch(nums, target, true)
  const r = binarySearch(nums, target, false) - 1
  if (l <= r && r < nums.length && nums[l] === target && nums[r] === target) {
    ans = [l, r]
  }
  return ans
}

let nums = [5, 7, 7, 8, 8, 10]
let target = 8
console.log(searchRange(nums, target))
